File: uikit/src/main/java/com/sendbird/uikit/widgets/EmojiReactionListView.java
package com.sendbird.uikit.widgets;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;

import com.sendbird.android.Reaction;
import com.sendbird.uikit.R;
import com.sendbird.uikit.activities.adapter.EmojiReactionListAdapter;
import com.sendbird.uikit.databinding.SbViewEmojiReactionListBinding;
import com.sendbird.uikit.interfaces.OnItemClickListener;
import com.sendbird.uikit.interfaces.OnItemLongClickListener;

import java.util.List;

public class EmojiReactionListView extends FrameLayout {
    private SbViewEmojiReactionListBinding binding;
    private EmojiReactionListAdapter adapter;
    private GridLayoutManager layoutManager;
    private final static int MAX_SPAN_SIZE = 4;

    public EmojiReactionListView(@NonNull Context context) {
        this(context, null);
    }

    public EmojiReactionListView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, R.attr.sb_widget_emoji_message);
    }

    public EmojiReactionListView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        binding = SbViewEmojiReactionListBinding.inflate(LayoutInflater.from(getContext()), this, true);
        binding.rvEmojiReactionList.setUseDivider(false);
        layoutManager = new GridLayoutManager(getContext(), MAX_SPAN_SIZE);
        binding.rvEmojiReactionList.setLayoutManager(layoutManager);
        binding.rvEmojiReactionList.setHasFixedSize(true);
        adapter = new EmojiReactionListAdapter();
        binding.rvEmojiReactionList.setAdapter(adapter);
    }

    public void setReactionList(@NonNull List<Reaction> reactionList) {
        if (adapter != null) {
            adapter.setReactionList(reactionList);
            resetSpanSize();
        }
    }

    private void resetSpanSize() {
        int itemSize = adapter.getItemCount();
        if (itemSize > 0) {
            layoutManager.setSpanCount(Math.min(itemSize, MAX_SPAN_SIZE));
        }
    }

    public void setEmojiReactionClickListener(@Nullable OnItemClickListener<String> emojiReactionClickListener) {
        if (adapter != null) {
            adapter.setEmojiReactionClickListener(emojiReactionClickListener);
        }
    }

    public void setEmojiReactionLongClickListener(@Nullable OnItemLongClickListener<String> emojiReactionLongClickListener) {
        if (adapter != null) {
            adapter.setEmojiReactionLongClickListener(emojiReactionLongClickListener);
        }
    }

    public void setMoreButtonClickListener(@Nullable OnClickListener moreButtonClickListener) {
        if (adapter != null) {
            adapter.setMoreButtonClickListener(moreButtonClickListener);
        }
    }

    public void setUseMoreButton(boolean useMoreButton) {
        if (adapter != null) {
            adapter.setUseMoreButton(useMoreButton);
        }
    }

    public boolean useMoreButton() {
        if (adapter != null) {
            return adapter.useMoreButton();
        }
        return true;
    }

    public void refresh() {
        if (adapter != null) {
            resetSpanSize();
            adapter.notifyItemRangeChanged(0, adapter.getItemCount());
        }
    }

    @Override
    public void setClickable(boolean clickable) {
        super.setClickable(clickable);
        if (adapter != null) {
            adapter.setClickable(clickable);
        }
    }

    @Override
    public void setLongClickable(boolean clickable) {
        super.setLongClickable(clickable);
        if (adapter != null) {
            adapter.setClickable(clickable);
        }
    }

    @NonNull
    public EmojiReactionListView getLayout() {
        return this;
    }

    @NonNull
    public SbViewEmojiReactionListBinding getBinding() {
        return binding;
    }
}

File: uikit/src/main/java/com/sendbird/uikit/internal/ui/reactions/EmojiReactionListView.kt
package com.sendbird.uikit.internal.ui.reactions

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.FrameLayout
import androidx.recyclerview.widget.GridLayoutManager
import com.sendbird.android.message.Emoji
import com.sendbird.android.message.Reaction
import com.sendbird.uikit.activities.adapter.EmojiReactionListAdapter
import com.sendbird.uikit.databinding.SbViewEmojiReactionListBinding
import com.sendbird.uikit.interfaces.OnItemClickListener
import com.sendbird.uikit.interfaces.OnItemLongClickListener
import com.sendbird.uikit.model.EmojiManager
import kotlin.math.min

internal class EmojiReactionListView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
) : FrameLayout(context, attrs) {
    val binding: SbViewEmojiReactionListBinding
    val layout: EmojiReactionListView
        get() = this

    private val adapter: EmojiReactionListAdapter
    private val layoutManager: GridLayoutManager
    var maxSpanSize: Int = 4

    init {
        binding = SbViewEmojiReactionListBinding.inflate(LayoutInflater.from(context), this, true)
        binding.rvEmojiReactionList.setUseDivider(false)
        layoutManager = GridLayoutManager(context, maxSpanSize)
        binding.rvEmojiReactionList.layoutManager = layoutManager
        binding.rvEmojiReactionList.setHasFixedSize(true)
        adapter = EmojiReactionListAdapter()
        binding.rvEmojiReactionList.adapter = adapter
    }

    fun setReactionList(reactionList: List<Reaction?>, totalEmojiList: List<Emoji> = EmojiManager.allEmojis) {
        adapter.setTotalEmojiList(totalEmojiList)
        adapter.setReactionList(reactionList.filterNotNull())
        resetSpanSize()
    }

    private fun resetSpanSize() {
        val itemSize = adapter.itemCount
        if (itemSize > 0) {
            layoutManager.spanCount = min(itemSize, maxSpanSize)
        }
    }

    fun setEmojiReactionClickListener(emojiReactionClickListener: OnItemClickListener<String>?) {
        adapter.setEmojiReactionClickListener(emojiReactionClickListener)
    }

    fun setEmojiReactionLongClickListener(emojiReactionLongClickListener: OnItemLongClickListener<String>?) {
        adapter.setEmojiReactionLongClickListener(emojiReactionLongClickListener)
    }

    fun setMoreButtonClickListener(moreButtonClickListener: OnClickListener?) {
        adapter.setMoreButtonClickListener(moreButtonClickListener)
    }

    internal fun setClickListeners(
        emojiReactionClickListener: OnItemClickListener<String>?,
        emojiReactionLongClickListener: OnItemLongClickListener<String>?,
        moreButtonClickListener: OnClickListener?
    ) {
        setEmojiReactionClickListener(emojiReactionClickListener)
        setEmojiReactionLongClickListener(emojiReactionLongClickListener)
        setMoreButtonClickListener(moreButtonClickListener)
    }

    fun setUseMoreButton(useMoreButton: Boolean) {
        adapter.setUseMoreButton(useMoreButton)
    }

    fun useMoreButton(): Boolean {
        return adapter.useMoreButton()
    }

    fun refresh() {
        resetSpanSize()
        adapter.notifyItemRangeChanged(0, adapter.itemCount)
    }

    override fun setClickable(clickable: Boolean) {
        super.setClickable(clickable)
        adapter.setClickable(clickable)
    }

    override fun setLongClickable(clickable: Boolean) {
        super.setLongClickable(clickable)
        adapter.setClickable(clickable)
    }
}
